### PR TITLE
[Merged by Bors] - refactor(SetTheory/ZFC/Basic): redefine `Definable`

### PR DESCRIPTION
**Pull request by astrainfinita**
This allows automatic synthesis using typeclasses, see `ZFSet.map`. The new definition is `ZFSet.Definable`. The old `PSet.Definable` has been deprecated.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/SetTheory/ZFC/Basic.lean [hunk 1]
```diff
@@ -512,11 +513,107 @@ end PSet
...
+abbrev Definable₁.out (f : ZFSet.{u} → ZFSet.{u}) [Definable₁ f] :
+    PSet.{u} → PSet.{u} :=
+  fun x ↦ Definable.out (fun s ↦ f (s 0)) ![x]
+
+lemma Definable₁.mk_out {f : ZFSet.{u} → ZFSet.{u}} [Definable₁ f]
+    {x : PSet} :
+    .mk (out f x) = f (.mk x) :=
+  Definable.mk_out ![x]
+
+/-- An abbrev of `ZFSet.Definable` for binary function. -/
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Basic.lean:560]:
```suggestion
/-- An abbrev of `ZFSet.Definable` for binary functions. -/
```

**Review by vihdzp — APPROVED** [review]
This is so much nicer, thanks!

**Comment by vihdzp** [comment]
Is there any reason to keep `Function.OfArity` after this? It seems very clear that using `Fin n → α → β` is strictly better.

File: Mathlib/SetTheory/ZFC/Basic.lean [hunk 2]
```diff
@@ -512,11 +513,108 @@ end PSet
...
+  Quotient.out_eq
+
+/-- A set function is "definable" if it is the image of some n-ary `PSet`
+  function. This isn't exactly definability, but is useful as a sufficient
+  condition for functions that have a computable image. -/
+class Definable (n) (f : (Fin n → ZFSet.{u}) → ZFSet.{u}) where
+  /-- Turns a definable function into an n-ary `PSet` function. -/
+  out : (Fin n → PSet.{u}) → PSet.{u}
+  /-- A set function `f` is the image of `Definable.out f`. -/
+  mk_out : ∀ xs, mk (out xs) = f (mk <| xs ·) := by simp
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Basic.lean:537]:
```suggestion
  mk_out xs : mk (out xs) = f (mk <| xs ·) := by simp
```

**Comment by YaelDillies** [comment]
> Is there any reason to keep `Function.OfArity` after this? It seems very clear that using `Fin n → α → β` is strictly better.

You haven't replied to the above

File: Mathlib/SetTheory/ZFC/Basic.lean [hunk 3]
```diff
@@ -509,11 +510,108 @@ end PSet
...
+  fun x ↦ Definable.out (fun s ↦ f (s 0)) ![x]
+
+lemma Definable₁.mk_out {f : ZFSet.{u} → ZFSet.{u}} [Definable₁ f]
+    {x : PSet} :
+    .mk (out f x) = f (.mk x) :=
+  Definable.mk_out ![x]
+
+/-- An abbrev of `ZFSet.Definable` for binary functions. -/
+abbrev Definable₂ (f : ZFSet.{u} → ZFSet.{u} → ZFSet.{u}) := Definable 2 (fun s ↦ f (s 0) (s 1))
+
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Basic.lean:560]:
```suggestion
```

**Comment by vihdzp** [comment]
Alright, I think the future of `Arity` is a question best saved for Zulip.

One last question: do you think this API could be useful for any other types? From what I understand, this notion of "definability" really only boils down to being able to provide a computable definition on the non-quotient types. If there's some other type where this is a relevant issue, then perhaps we could move out the `Definable` API into its own file.

Otherwise, LGTM.

**Review by vihdzp — APPROVED** [review]

**Comment by YaelDillies** [comment]
Thanks! 🚀

maintainer merge

File: Mathlib/SetTheory/ZFC/Basic.lean (hunk 3 above)
Comment (vihdzp) [reply Mathlib/SetTheory/ZFC/Basic.lean:560]: Oh I'm sorry, mobile line-wrapping had me think there were two spaces here instead of one.

**Comment by fpvandoorn** [comment]
Sorry for the delay, and thanks!

bors merge

**Comment by YaelDillies** [comment]
You need to merge master

**Pull request closed without merging**